### PR TITLE
Fix: querying rollup data does not show aggregated points corresponding to points in ROB

### PR DESCRIPTION
**Pull request by zlin179**
Fix for issue https://github.com/grafana/metrictank/issues/1485

**Describe your changes**
The rollup aggregators only get the points once they leave the ROB.
This PR fixs aggregator so when being queried it will return the current value (considering points in ROB) but will not actually consume the points in ROB.

**Testing performed**
* Add some unit tests for it and pass it.
* Integrate test by comparing raw data and rollup data queried at the same time

mt-fix

### CODE REVIEW

File: mdata/aggregator.go [hunk 1]
```diff
@@ -99,6 +100,38 @@ func (agg *Aggregator) flush() {
 	agg.agg.Reset()
 }
 
+// Foresee duplicates the underlying Aggregation,
+// consumes points in the future, and return the newly aggregated points.
+// Foresee won't actually add those points to AggMetric.
+func (agg *Aggregator) Foresee(consolidator consolidation.Consolidator, from, to uint32, futurePoints []schema.Point) []schema.Point {
+	aggregationPreview := *(agg.agg)
```
Comment (Dieterbe) [inline mdata/aggregator.go:107]: wow this is clever. so we make a copy of the aggregator, scoped to only this request, to temporarily add the ROB's points to, so we can compute the full aggregation. nice... i could not come up with this.

File: mdata/aggregator.go [hunk 2]
```diff
@@ -99,6 +100,38 @@ func (agg *Aggregator) flush() {
...
+			if aggregationPreview.Cnt != 0 {
+				value, _ := aggregationPreview.GetValueFor(consolidator)
+				aggregatedPoints = append(aggregatedPoints, schema.Point{Val: value, Ts: currentBoundary})
+				aggregationPreview.Reset()
+			}
+			currentBoundary = boundary
+		}
+		aggregationPreview.Add(point.Val)
+	}
+	if aggregationPreview.Cnt != 0 && currentBoundary <= to {
```
Comment (Dieterbe) [inline mdata/aggregator.go:128]: i think to is exclusive, like everywhere else

File: mdata/aggmetric.go [hunk 3]
```diff
@@ -163,25 +163,38 @@ func (a *AggMetric) GetAggregated(consolidator consolidation.Consolidator, aggSp
...
+			result, err := agg.Get(from, to)
+			if err != nil {
+				return Result{}, err
 			}
-			return agg.Get(from, to)
+			futurePoints := []schema.Point{}
+			if a.rob != nil {
+				futurePoints = a.rob.Get()
+				if len(futurePoints) > 0 && result.Oldest > futurePoints[0].Ts {
+					result.Oldest = futurePoints[0].Ts
```
Comment (Dieterbe) [inline mdata/aggmetric.go:192]: why? this is definitely not obvious and needs a comment.
But I disagree here.
In Foresee, we disregard all ROB samples that correspond to a boundary preceeding the aggregator boundary. and any sample read from agg.Get() would be one that has been flushed by the aggregator, ie samples that have a boundary lower than the boundary in the aggregator.

Thus, it follows that we wouldn't use a sample from futurePoints that is older than one present in result. Thus I don't think we should change the Oldest value here.

Comment (shanson7) [reply mdata/aggmetric.go:192]: What if all of the datapoints for this metric are in the `rob`?

Comment (Dieterbe) [reply mdata/aggmetric.go:192]: in that case, the adjustment makes sense.
I think we can make it work for both cases by only doing the adjustment when oldest is maxInt32 (kind of a hack), or better when `len(result.Iters) == 0`

File: mdata/aggmetric_test.go [hunk 4]
```diff
@@ -486,14 +486,13 @@ func TestGetAggregated(t *testing.T) {
 	m.Add(30, 30)
 	m.Add(31, 31)
 	m.Add(32, 32)
-	m.Add(40, 40)
```
Comment (Dieterbe) [inline mdata/aggmetric_test.go:489]: why is this removed?

File: mdata/aggmetric.go [hunk 5]
```diff
@@ -163,25 +163,38 @@ func (a *AggMetric) GetAggregated(consolidator consolidation.Consolidator, aggSp
...
-				return Result{}, fmt.Errorf("Consolidator %q not configured", consolidator)
+				return Result{}, fmt.Errorf("consolidator %q not configured", consolidator)
+			}
+			result, err := agg.Get(from, to)
+			if err != nil {
+				return Result{}, err
 			}
-			return agg.Get(from, to)
+			futurePoints := []schema.Point{}
+			if a.rob != nil {
```
[the remark this answers is not in the document]
Comment (zlin179) [reply mdata/aggmetric.go:189]: I think `a.RLock()` may be enough like line 226 in `AggMetric.Get()` as it's just reading?
Not quite sure about the `aggregator` as there's a comment above `// no lock needed cause aggregators don't change at runtime`. If it's causing problems it may be another issue? To my understanding `aggregator.xxxMetric` has its own lock so it should be fine.

File: mdata/aggmetric_test.go (hunk 4 above)
Comment (zlin179) [reply mdata/aggmetric_test.go:489]: I removed it to test the bug fix. The span is 5 and 40 is the boundary. If (40, 40) is there it behaves the same (aggregator will flush and `len(futurePoints)==0`).
I can also add extra points like m.Add(41, 41) m.Add(42, 42) to test the 45 boundary but I think they are kind of the same. I can update this if you prefer.

File: mdata/aggmetric.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Dieterbe) [reply mdata/aggmetric.go:189]: If it's just for the rob, then indeed RLock will suffice, not Lock.

That comment refers to accessing a.aggregators itself - the slice with pointers.  But I'm talking about the actual aggregators being pointed to. I think it has a few fields that we unsafely read and write. It's hard to believe we would have had such an obvious bug for years, so let me look into it .

File: mdata/aggmetric_test.go (hunk 4 above)
Comment (zlin179) [reply mdata/aggmetric_test.go:489]: I removed it to test the bug fix. The span is 5 and 40 is the boundary. If (40, 40) is there it behaves the same (aggregator will flush to the underlying aggMetric).
So if (40, 40) is removed, before this PR, ({31+32, Ts:35}) will not be in the result.
I can also add extra points like `m.Add(41, 41)` `m.Add(42, 42)` to test the 45 boundary but I think they are kind of the same. I can update this if you prefer.

Comment (Dieterbe) [reply mdata/aggmetric_test.go:489]: makes sense, what you have.

File: mdata/aggmetric.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Dieterbe) [reply mdata/aggmetric.go:189]: i think what we need is:
* https://github.com/grafana/metrictank/pull/2025, this should also clarify things
* a RLock here around the rob acess and the Foresee() call.
